caramel/compiler/reason_to_erlang/reason_to_erlang.ml

```diff
@@ -0,0 +1,66 @@
+open Compile_common
+
+let tool_name = "caramel:re-to-erl"
+
+let setup_lexbuf ~parser source_file =
+  let ic = open_in source_file in
+  try
+    seek_in ic 0;
+    let lexbuf = Lexing.from_channel ic in
+    Location.init lexbuf source_file;
+    parser lexbuf
+  with e ->
+    Format.fprintf Format.std_formatter "Reason parser error: %s"
+      (Printexc.to_string e);
+    close_in_noerr ic;
+    exit 1
+
+let read_signature info =
+  let module_name = info.module_name in
+  let cmi_file = module_name ^ ".cmi" in
+  try
+    let intf_file = Load_path.find_uncap cmi_file in
+    let sign = Env.read_signature module_name intf_file in
+    Some sign
+  with Not_found -> None
+
+let parse_implementation source_file =
+  let implementation lexbuf =
+    lexbuf |> Reason_toolchain.RE.implementation
+    |> Reason_toolchain.To_current.copy_structure
+  in
+  setup_lexbuf ~parser:implementation source_file
+
+let parse_interface source_file =
+  let interface lexbuf =
+    lexbuf |> Reason_toolchain.RE.interface
+    |> Reason_toolchain.To_current.copy_signature
+  in
+  setup_lexbuf ~parser:interface source_file
+
+let interface ~source_file ~output_prefix ~opts:_ =
+  Compile_common.with_info ~native:false ~tool_name ~source_file ~output_prefix
+    ~dump_ext:"cmo"
+  @@ fun _info ->
+  let _parsetree = parse_interface source_file in
```

Comment:
`Optcompile.interface` lives here:

https://github.com/ocaml/ocaml/blob/trunk/driver/optcompile.ml#L26-L28

And ends up calling `Compile_common.interface`, which lives here:

https://github.com/ocaml/ocaml/blob/trunk/driver/compile_common.ml#L84-L92

But the args of this function expect a file. If we look inside it you'll see there's a single call to parse the file that returns a parsetree (`ast`) and then it does this:

```ocaml
    let tsg = typecheck_intf info ast in
    if not !Clflags.print_types then begin
      emit_signature info ast tsg
    end
```

I'm afraid some of those functions aren't exported, but we can replicate them in here since they rely on other Typemod functions that are in fact public.

Reply:
Hi Leandro!

Thanks for your feedbacks! It does help a lot! :)